---
  output: html_document
---
  
##### Jongchul Seon

##### DNA Methylation example (Bisulfite Sequencing, Illumina 450k array)

------------------------

##### Load 

  
##### We are going to use public data,from http://www.ncbi.nlm.nih.gov/gds/?term=GSE68777. You can check The journal and main author here,http://www.ncbi.nlm.nih.gov/pubmed/26147665 and http://kasperdanielhansen.github.io/genbioconductor/html/minfi.html


##### Microarray-based methods to determine pattern of methylation. The Illumina Methylation Assay is one such assay that applies the bisulfite sequencing technology on a microarray level to generate genome-wide methylation data.

##### We are going to compare DNA Methylation levels between two groups, Acute Mania(20) <-> Control(20)


```{r chunk1,echo=TRUE}

library(devtools)
library(GEOquery)
library(minfi)
library(doParallel) ##CRAN

library(GenomicRanges)
library(doParallel)

library(devtools)

##library(pkgmaker)
##library(minfiData)
##library(AnnotationHub)
##library(epivizr)


## source("http://www.bioconductor.org/biocLite.R")
## biocLite(c("minfi"))

## getwd()

## data download  

#getGEOSuppFiles("GSE68777")
#untar("GSE68777/GSE68777_RAW.tar", exdir = "GSE68777/idat")
#head(list.files("GSE68777/idat", pattern = "idat"))

#idatFiles <- list.files("GSE68777/idat", pattern = "idat.gz$", full = TRUE)

#sapply(idatFiles, gunzip, overwrite = TRUE)

## data read

rgSet <- read.450k.exp("GSE68777/idat")
rgSet


```


-------------------------

##### idat is raw data file format. 


```{r chunk2,echo=TRUE}

green <- getGreen(rgSet)
red <- getRed(rgSet)

par(mfrow=c(1,2))

plot(rowMeans(red[1:500,]))
plot(rowMeans(green[1:500,]))

par(mfrow=c(1,1))
plot(rowMeans(red)-rowMeans(green))

par(mfrow=c(1,2))
hist(green[1,])
hist(red[1,])

## probe locations can be devided into 4 categories of CpG island. (Island, OpenSea, Shelf, Shore)
table(getIslandStatus(rgSet))

pData(rgSet)

##class(rgSet)
##??RGChannelSet
##??minfi

## library(illuminaio)
## library(IlluminaDataTestFiles)

## head(list.files("GSE68777/idat", pattern = "idat"))

## green1 <- readIDAT("D:/R/GSE68777/idat/GSM1681154_5958091019_R03C02_Grn.idat")

## red1 <- readIDAT("D:/R/GSE68777/idat/GSM1681154_5958091019_R03C02_Red.idat")

## names(green1)
## head(green1$Quants)

##summary(green1$Quants[,3])
##summary(red1$Quants[,1])
##hist(log2(green1$Quants[,1]))
##hist(log2(red1$Quants[,1]))


##tempdir()


```


##### The phenotype data 


```{r chunk3,echo=TRUE}

###getGEO
###geoMat <- getGEO("GSE68777")
### destdir = "D:\\R" => window OS directory

geoMat <- getGEO("GSE68777",destdir = "D:\\R")

pD.all <- pData(geoMat[[1]])

class(pD.all)
head(pD.all)
names(pD.all)

rgSet


```


##### The phenoData(pD) and the featureData(rgSet) are modified for further analysis.

```{r chunk4,echo=TRUE}



pD <- pD.all[, c("title", "geo_accession", "characteristics_ch1.1", "characteristics_ch1.2")]
head(pD)

names(pD)[c(3,4)] <- c("group", "sex")
head(pD)

##pD$group
pD$group <- sub("^diagnosis: ", "", pD$group)
##pD$group

##pD$sex 
pD$sex <- sub("^Sex: ", "", pD$sex)
##pD$sex

##sampleNames(rgSet) 
sampleNames(rgSet) <- sub(".*_5", "5", sampleNames(rgSet))
##sampleNames(rgSet) 

rownames(pD) <- pD$title
pD <- pD[sampleNames(rgSet),]

head(pD)


pData(rgSet) <- pD
head(pData(rgSet))
rgSet


```


##### Beta = Meth / (Meth + Unmeth + offset)

##### Beta value of Each sample should be left and right skewed.

```{r chunk5,echo=TRUE}

par(mfrow=c(1,1))

densityPlot(rgSet, sampGroups = pD$group, main = "Beta", xlab = "Beta")

par(oma=c(2,10,1,1))
densityBeanPlot(rgSet, sampGroups = pD$group,sampNames = pD$title)

```


##### Normalization 

##### Beta = M/(M + U + 100) => Illumina formula

##### M-values are perhaps an unfortunate terminology, but it seems to be standard in the methylation array world. These are computed as logit(Beta) and are obtained by getM.

```{r chunk6,echo=TRUE}

MSet.raw <- preprocessRaw(rgSet)

getMeth(MSet.raw)[1:4,1:3]

getUnmeth(MSet.raw)[1:4,1:3]

getBeta(MSet.raw, type = "Illumina")[1:4,1:3]

getM(MSet.raw)[1:4,1:3]

MSet.norm <- preprocessIllumina(rgSet, bg.correct = TRUE, normalize = "controls", reference = 2)

```


##### Finding diferentially methylated positions (DMPs)

```{r chunk7,echo=TRUE}

M <- getM(MSet.norm[1:20000,], type = "beta", betaThreshold = 0.001)

dmp <- dmpFinder(M, pheno=pD$group, type="categorical")

head(dmp)

cpgs <- rownames(dmp)[1:4]

par(mfrow=c(2,2))

plotCpg(MSet.norm[1:20000,], cpg=cpgs, pheno=pD$group)

```



##### Finding diferentially methylated positions (Bumphunting)


```{r chunk8,echo=TRUE}

dat = preprocessIllumina(rgSet)
dat = mapToGenome(dat)

dat = ratioConvert(dat,type="Illumina")

detectCores()

registerDoParallel(cores = 4)


group =pData(dat)$group
X= model.matrix(~group)
index = which(seqnames(dat)=="chr15")

dat = dat[index,] ## for illustrative purposes
res=bumphunter(dat,X,cutoff=0.1,B=1000)

res

table(getIslandStatus(dat))


```


```{r chunk9,echo=TRUE}

devtools::session_info()

```
