---
output: pdf_document
---

##### Jongchul Sun

##### RNA Count example 1 

------------------------

##### Load 

##### biological replicates C57BL/6J (10) <-> DBA/2J (11) are hybridized in each experiment and lane.

##### experiment number and lane.number are different.There will be some variances among these. 

##### We are going to use limma package to compare  the means of gene expression values for two groups of replicates for a given gene



```{r chunk1,echo=TRUE}
library(devtools)
library(Biobase)
library(goseq)
library(limma)
library(genefilter)
#### library(DESeq2)


con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bottomly_eset.RData")
load(file=con)
close(con)
bot = bottomly.eset
pdata_bot=pData(bot)
fdata_bot = featureData(bot)
edata = exprs(bot)

sum(is.na(edata))



dim(edata)
dim(pdata_bot)
dim(fdata_bot)
head(edata)
head(pdata_bot)

unique(pdata_bot[,1])

```


---------------- 

##### Genes whose average counts are over than 5 are selected and log(2) transformed.
##### A common pre-processing technique is to remove features that don't have much data


```{r chunk2,echo=TRUE}

par(mfrow=c(1,2))

boxplot(edata,col=2,range=0)
boxplot(log2(edata+1),col=2,range=0)

hist(log2(edata[,1]+1),col=2)
hist(log2(edata[,2]+1),col=2)

mm = log2(edata[,1]+1) - log2(edata[,2]+1)
aa = log2(edata[,1]+1) + log2(edata[,2]+1)
plot(aa,mm,col=2)


fdata_bot = fdata_bot[rowMeans(edata) > 5]
edata = edata[rowMeans(edata) > 5, ]
edata = log2(as.matrix(edata) + 1)

hist(edata,,col=2,range=0)


```

-----------------------

##### NCBI Build 37, mm9. check out the paper which describes the data, Evaluating Gene Expression in C57BL/6J and DBA/2JMouse Striatum Using RNA-Seq and Microarrays.  http://www.ncbi.nlm.nih.gov/pubmed?term=21455293. 


```{r chunk3,echo=TRUE}

class(supportedGenomes())
isS4(supportedGenomes())

head(supportedGenomes())

species <- supportedGenomes()[,2]
## species

## source("http://www.bioconductor.org/biocLite.R")
## biocLite("org.Mm.eg.db")


#### species[species %in% "Mouse"]


```


--------------------------

##### Using limma package to find differently expressed genes between two strands. 


```{r chunk4,echo=TRUE}

mod = model.matrix(~ pdata_bot$strain)
fit_limma = lmFit(edata,mod)
ebayes_limma = eBayes(fit_limma)


```


-------------------------------------

##### limma dose moderated t-statistics. Moderated t-statistics lead to pvalues in the same way that ordinary t-statistics do except that the degrees of freedom are increased, reflecting the greater reliability associated with the smoothed standard errors.


```{r chunk5,echo=TRUE}

par(mfrow=c(1,1))

tstats_obj = rowttests(edata,pdata_bot$strain)
##names(tstats_obj)
hist(tstats_obj$statistic,col=2)

plot(ebayes_limma$t[,2],tstats_obj$statistic,col=4,
     xlab="Moderated T-stat",ylab="T-stat")


```


----------------

###### Bonferroni and Benjamini-Hochberg FDR correction with p.adjust

###### You can use the `p.adjust` function to get "multiple testing corrected" p-values which you can then use to control error rates.


##### 223 genes are differently expressesed between two strands at the 5% FDR level using Benjamini-Hochberg correction.


```{r chunk6,echo=TRUE}

limma_pvals = topTable(ebayes_limma,adjust.method="BH",sort.by="none",number=dim(edata)[1])

dim(limma_pvals)

sum(limma_pvals$adj.P.Val < 0.05)

genes <- limma_pvals$adj.P.Val < 0.05

DE <- featureNames(fdata_bot)[genes]

length(DE)

```

--------

##### Gene Set analysis using goseq package
##### using database of 'Gene Ontology Consortium' to do gene set analysis


```{r chunk7,echo=TRUE}

genes = as.integer(limma_pvals$adj.P.Val < 0.05)
not_na = !is.na(genes)
names(genes) = rownames(edata)
##names(genes)

genes = genes[not_na]

head(supportedGenomes(),n=12)[,1:4]

??nullp

pwf=nullp(genes,"mm9","ensGene")
head(pwf)


GO.MF=goseq(pwf,"mm9","ensGene",test.cats=c("GO:MF"))
head(GO.MF)

GO.MF[1:10,]

```



```{r chunk8,echo=TRUE}

devtools::session_info()

``````
